File: solutions/4096-sort-array-using-prefix-reversals/sort-array-using-prefix-reversals.py
from collections import deque
from typing import List

class Solution:
    def sortArray(self, nums: List[int], pre: List[int]) -> int:
        start = tuple(nums)
        target = tuple(sorted(nums))
        
        if start == target:
            return 0
            
        queue = deque([start])
        visited = set([start])
        res = 0
        
        while queue:
            for _ in range(len(queue)):
                curr = queue.popleft()
                
                for p in pre:
                    # Slice and reverse the prefix, then append the rest
                    next_state = curr[:p][::-1] + curr[p:]
                    
                    if next_state == target:
                        return res + 1
                        
                    if next_state not in visited:
                        visited.add(next_state)
                        queue.append(next_state)
                        
            res += 1
            
        return -1

File: solutions/4096-sort-array-using-prefix-reversals/sort-array-using-prefix-reversals.cpp
#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>

using namespace std;

// Custom hash function for vector<int> to be used in unordered_set
struct VectorHasher {
    size_t operator()(const vector<int>& V) const {
        size_t hash = V.size();
        for(auto& i : V) {
            hash ^= i + 0x9e3779b9 + (hash << 6) + (hash >> 2);
        }
        return hash;
    }
};

class Solution {
public:
    int sortArray(vector<int>& nums, vector<int>& pre) {
        vector<int> start = nums;
        vector<int> target = nums;
        sort(target.begin(), target.end());
        
        if (start == target) return 0;
        
        queue<vector<int>> q;
        unordered_set<vector<int>, VectorHasher> visited;
        
        q.push(start);
        visited.insert(start);
        
        int res = 0;
        
        while (!q.empty()) {
            int size = q.size();
            for (int j = 0; j < size; j++) {
                vector<int> curr = q.front();
                q.pop();
                
                for (int p : pre) {
                    vector<int> nextState = curr;
                    // Reverse the prefix
                    reverse(nextState.begin(), nextState.begin() + p);
                    
                    if (nextState == target) {
                        return res + 1;
                    }
                    
                    if (visited.find(nextState) == visited.end()) {
                        visited.insert(nextState);
                        q.push(nextState);
                    }
                }
            }
            res++;
        }
        
        return -1;
    }
};